packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,381 @@
+    if (!instance)
+      instance = new ConnectionManager(
+        libp2p,
+        keepAliveOptions,
+        relay,
+        options
+      );
+    ConnectionManager.instances.set(peerId, instance);
+
+    return instance;
+  }
+
+  constructor(
+    libp2pComponents: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ) {
+    this.libp2pComponents = libp2pComponents;
+    this.options = options ?? {};
+
+    this.runService(keepAliveOptions, relay)
+      .then(() => log(`Connection Manager is now running`))
+      .catch((error) => log(`Unexpected error while running service`, error));
+  }
+
+  private async runService(
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay
+  ): Promise<void> {
+    const { loggingIntervalMs = DEFAULT_LOGGING_INTERVAL_MS } = this.options;
+
+    // log state every N seconds
+    setInterval(async () => {
+      this.fetchUpdatedState();
+    }, loggingIntervalMs);
+
+    // start event listeners
+    this.startPeerDiscoveryListener();
+    this.startPeerConnectionListener(keepAliveOptions, relay);
+    this.startPeerDisconnectionListener();
+  }
+
+  stopService(): void {
+    this.keepAliveManager.stopAll();
+    this.libp2pComponents.connectionManager.removeEventListener("peer:connect");
+    this.libp2pComponents.connectionManager.removeEventListener(
+      "peer:disconnect"
+    );
+    this.libp2pComponents.peerStore.removeEventListener("peer");
+  }
+
+  private async dialPeer(peerId: PeerId): Promise<void> {
+    const { maxDialAttemptsForPeer = DEFAULT_MAX_DIAL_ATTEMPTS_FOR_PEER } =
+      this.options;
+
+    try {
+      log(`Dialing peer ${peerId.toString()}`);
+      await this.libp2pComponents.dial(peerId);
+
+      const tags = await this.getTagNamesForPeer(peerId);
+      // add tag to connection describing discovery mechanism
+      this.libp2pComponents.connectionManager
+        .getConnections(peerId)
+        .forEach((conn) => conn.tags.push(...tags));
+
+      this.dialAttemptsForPeer.delete(peerId.toString());
+    } catch (error) {
+      const dialAttempt = this.dialAttemptsForPeer.get(peerId.toString()) ?? 1;
+      this.dialAttemptsForPeer.set(peerId.toString(), dialAttempt + 1);
+
+      //remove peer from peerStore if dial fails after max attempts
+      if (dialAttempt > maxDialAttemptsForPeer) {
+        try {
+          log(
+            "Deleting undialable peer " +
+              peerId.toString() +
+              " from peer store as max dial attempts reached"
+          );
+          await this.libp2pComponents.peerStore.delete(peerId);
+          return;
+        } catch (error) {
+          throw (
+            "Error deleting undialable peer" +
+            peerId.toString() +
+            " from peerStore"
+          );
+        }
+      }
+
+      log(
+        `Failed to dial ${peerId.toString()} - reattempting (${dialAttempt})`
+      );
+
+      this.dialPeer(peerId);
```

Comment:
Is there a way to rewrite it in a linear way? I understand that you wanted to improve readability of this code, but I think it is easy to make a mistake and have an unintended loop.

Reply:
I wasn't thinking of anything in particular. 

We can rewrite it something like this (didn't test though):
```
let { maxDialAttemptsForPeer = DEFAULT_MAX_DIAL_ATTEMPTS_FOR_PEER } =
      this.options;

    let dialAttempt = 1;

    while (dialAttempt < maxDialAttemptsForPeer) {
      try {
        log(`Dialing peer ${peerId.toString()}`);

        const [_dial, tags] = await Promise.all([this.libp2pComponents.dial(peerId), this.getTagNamesForPeer(peerId)]);

        this.libp2pComponents.connectionManager
          .getConnections(peerId)
          .forEach((conn) => conn.tags.push(...tags));
        this.dialAttemptsForPeer.delete(peerId.toString());
        return;
      } catch(_error) {
        dialAttempt += 1;
      }
    }

    try {
      log(`Deleting undialable peer ${peerId.toString()} from peer store as max dial attempts reached`);
      await this.libp2pComponents.peerStore.delete(peerId);
    } catch (_error) {
      const message = `Error deleting undialable peer ${peerId.toString()} from peerStore`;
      log(message);
      throw Error(message);
    }
```

Or any other way that you think reasonable. It just seems to me that because of recursion code complexity is increased and you use `dialAttempts` map and try within try.